### PR TITLE
feat: Add support for custom refresh handling

### PR DESCRIPTION
**Pull request by blumewas**
### 🔗 Linked issue

Issue #714 

### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

The current behaviour for refreshing the users tokens does not provide many options for me as developer. As a developer I would like to be able to extend or customize this behaviour.

This PR:
- adds new types `RefreshHandler` and `RefreshHandlerConfig`
- adds the possibility to configure a custom `RefreshHandler` via `refreshHandler?` property to `auth.session`
- extracts the default refreshing behaviour to a separate file

### 📝 Checklist

- [x] I have linked an issue or discussion.
- [ ] I have added tests (if possible).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/runtime/plugin.ts [hunk 1]
```diff
@@ -23,61 +25,20 @@ export default defineNuxtPlugin(async (nuxtApp) => {
   }
 
   // 2. Setup session maintanence, e.g., auto refreshing or refreshing on foux
-  const { enableRefreshOnWindowFocus, enableRefreshPeriodically } =
-    runtimeConfig.session
+  // const RefreshHandler = runtimeConfig.session.refreshHandler ?? DefaultRefreshHandler
```
Comment (phoenix-ru) [inline src/runtime/plugin.ts:28]: could you please remove this?

File: src/runtime/utils/refreshHandler.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    this.config = config
+
+    const runtimeConfig = useRuntimeConfig().public.auth
+
+    const { data } = useAuthState()
+    const { getSession } = useAuth()
+
+    document.addEventListener('visibilitychange', () => {
+      this.visibilityHandler()
+    }, false)
```
Comment (phoenix-ru) [inline src/runtime/utils/refreshHandler.ts:39]: I think you changed the semantics here: you are adding an anonymous event handler and not really removing it on `destroy`

File: src/runtime/utils/refreshHandler.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        }
+      }, intervalTime)
+    }
+  },
+
+  destroy (): void {
+    // Clear visibility handler
+    document.removeEventListener('visibilitychange', () => {
+      this.visibilityHandler()
+    }, false)
```
Comment (phoenix-ru) [inline src/runtime/utils/refreshHandler.ts:69]: not actually removing the handler here

File: src/runtime/utils/refreshHandler.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
+import type { RefreshHandlerConfig, RefreshHandler } from '../types'
+import { useRuntimeConfig, useAuth, useAuthState } from '#imports'
+
+export const defaultRefreshHandler: RefreshHandler & {
```
Comment (phoenix-ru) [inline src/runtime/utils/refreshHandler.ts:4]: May be better to define an interface instead:
```ts
interface DefaultRefreshHandler extends RefreshHandler {
  // ...
}

const defaultRefreshHandler: DefaultRefreshHandler = {
```

Comment (phoenix-ru) [inline src/runtime/utils/refreshHandler.ts:4]: And use
```ts
export default defaultRefreshHandler
```

**Review by phoenix-ru — CHANGES_REQUESTED** [review]

**Comment by zoey-kaiser** [comment]
Hi @blumewas 👋 

Have you had a chance to look at the review yet? Otherwise are there any other open questions I can help you with?

**Comment by blumewas** [comment]
Hi @zoey-kaiser,
I will try to put it on my agenda for today/this evening. The review got completely lost in all other PRs from work. Sorry for that.

File: src/runtime/plugin.ts (hunk 1 above)
Comment (blumewas) [reply src/runtime/plugin.ts:28]: ✅

File: src/runtime/utils/refreshHandler.ts (hunk 4 above)
Comment (blumewas) [reply src/runtime/utils/refreshHandler.ts:4]: ✅

File: src/runtime/utils/refreshHandler.ts (hunk 2 above)
Comment (blumewas) [reply src/runtime/utils/refreshHandler.ts]: Done. Now using the function ref inside the object

File: src/runtime/utils/refreshHandler.ts (hunk 3 above)
Comment (blumewas) [reply src/runtime/utils/refreshHandler.ts]: Done. Now using the function ref inside the object

File: src/runtime/utils/refreshHandler.ts (hunk 4 above)
Comment (blumewas) [reply src/runtime/utils/refreshHandler.ts]: ✅

**Comment by blumewas** [comment]
Hey @zoey-kaiser,
I have implemented the requested changes made by @phoenix-ru. I have also merged the current main into my feature branch.

**Comment by blumewas** [comment]
Sorry my bad 😅 I have now removed the console statement which made the linter fail

**Review by phoenix-ru — APPROVED** [review]

**Comment by blumewas** [comment]
I can do that later today. If you want to merge it already then it would be better if you will do it 😄

**Pull request merged**